apis/apiextensions/v1beta1/composition_types_test.go

```diff
@@ -22,9 +22,203 @@ import (
 	"github.com/google/go-cmp/cmp"
 	"github.com/pkg/errors"
 
+	"k8s.io/utils/pointer"
+
 	"github.com/crossplane/crossplane-runtime/pkg/test"
 )
 
+func TestPatchTypeReplacement(t *testing.T) {
+	type args struct {
+		comp    CompositionSpec
+		patches []Patch
+	}
+	type want struct {
+		patches []Patch
+		err     error
+	}
+
+	cases := map[string]struct {
+		args
+		want
+	}{
+		"NoCompositionPatchTypes": {
+			args{
+				comp: CompositionSpec{},
+				patches: []Patch{
+					{
+						Type: PatchTypeFromCompositeFieldPath,
+						PatchFromCompositeFieldPath: &PatchFromCompositeFieldPath{
+							FromFieldPath: pointer.StringPtr("metadata.name"),
+						},
+					},
+					{
+						Type: PatchTypeFromCompositeFieldPath,
+						PatchFromCompositeFieldPath: &PatchFromCompositeFieldPath{
+							FromFieldPath: pointer.StringPtr("metadata.namespace"),
+						},
+					},
+				},
+			},
+			want{
+				patches: []Patch{
+					{
+						Type: PatchTypeFromCompositeFieldPath,
+						PatchFromCompositeFieldPath: &PatchFromCompositeFieldPath{
+							FromFieldPath: pointer.StringPtr("metadata.name"),
+						},
+					},
+					{
+						Type: PatchTypeFromCompositeFieldPath,
+						PatchFromCompositeFieldPath: &PatchFromCompositeFieldPath{
+							FromFieldPath: pointer.StringPtr("metadata.namespace"),
+						},
+					},
+				},
+			},
+		},
+		"UndefinedPatchSet": {
+			args{
+				comp: CompositionSpec{},
+				patches: []Patch{
+					{Type: PatchTypePatchSet, PatchSet: &PatchSet{ID: pointer.StringPtr("patch-set-1")}},
+				},
+			},
+			want{
+				// We expect the internal Patches field to be unmodified
```

Comment:
We have a convention (in some tests) of including a `reason` string in each test case that captures this kind of data, both as documentation for test readers and as something that can be printed if the test fails.